src/lib/ledger.rs

```diff
@@ -0,0 +1,349 @@
+use std::{
+    cell::Cell,
+    fmt,
+    sync::{Arc, Mutex, Weak},
+    time::Duration,
+};
+
+use anyhow::Context;
+use bip32::DerivationPath;
+use candid::Principal;
+use hidapi::HidApi;
+use ic_agent::{agent::EnvelopeContent, Identity, Signature};
+use indicatif::ProgressBar;
+use ledger_apdu::{APDUAnswer, APDUCommand, APDUErrorCode};
+use ledger_transport_hid::TransportNativeHID;
+use once_cell::sync::Lazy;
+use serde::Serialize;
+use serde_cbor::Serializer;
+
+use super::{
+    derivation_path, genesis_token_canister_id, governance_canister_id, ledger_canister_id,
+    AnyhowResult,
+};
+
+const CLA: u8 = 0x11;
+const GET_VERSION: u8 = 0x00;
+const GET_ADDR_SECP256K1: u8 = 0x01;
+const SIGN_SECP256K1: u8 = 0x02;
+const P1_ONLY_RETRIEVE: u8 = 0x00;
+const P1_SHOW_ADDRESS_IN_DEVICE: u8 = 0x01;
+const TX_NORMAL: u8 = 0x00;
+const TX_STAKING: u8 = 0x01;
+
+const PAYLOAD_INIT: u8 = 0x00;
+const PAYLOAD_ADD: u8 = 0x01;
+const PAYLOAD_LAST: u8 = 0x02;
+
+const PK_OFFSET: usize = 0;
+const PK_LEN: usize = 65;
+const PRINCIPAL_OFFSET: usize = PK_LEN;
+const PRINCIPAL_LEN: usize = 29;
+
+const SIG_OFFSET: usize = 43;
+const SIG_LEN: usize = 64;
+
+const CHUNK_SIZE: usize = 250;
+
+// necessary due to HidApi being a singleton
+static GLOBAL_HANDLE: Lazy<Mutex<Weak<LedgerIdentityInner>>> =
+    Lazy::new(|| Mutex::new(Weak::new()));
+
+thread_local! {
+    static NEXT_STAKE: Cell<bool> = Cell::new(false);
+}
+
+struct LedgerIdentityInner {
+    transport: Mutex<TransportNativeHID>,
+}
+
+pub struct LedgerIdentity {
+    inner: Arc<LedgerIdentityInner>,
+}
+
+impl LedgerIdentity {
+    pub fn new() -> AnyhowResult<Self> {
+        let mut global = GLOBAL_HANDLE.lock().unwrap();
+        if let Some(existing) = global.upgrade() {
+            Ok(Self { inner: existing })
+        } else {
+            let inner = Arc::new(LedgerIdentityInner {
+                transport: Mutex::new(TransportNativeHID::new(&HidApi::new().unwrap())?),
+            });
+            *global = Arc::downgrade(&inner);
+            Ok(Self { inner })
+        }
```

Comment:
From an organizational perspective, I'd suggest moving some of this logic into `LedgerIdentityInner`.  Here, something like
```suggestion
        let inner = LedgerIdentityInner::get_or_create()?;
        Ok(Self { inner })
```

Similarly, all of the code that uses `self.inner.transport` would instead call `self.inner.something()`.  All the spinner stuff would go in one place, too.  I left some suggestions below for how this might look, but I'm sure you get the idea.

Reply:
I'm not sure I agree with this change - the code is currently in two pieces, a series of functions that talk to a USB device over a handle and an abstraction that adapts it to the Identity API including multi-instantiation. I don't think adding a third layer, or enmeshing the second layer in the abstraction, will make the code organizationally cleaner. Perhaps it could be a module.